text/0083-rawkv-cross-cluster-replication.md

```diff
@@ -0,0 +1,185 @@
+
+Besides, **TiKV-CDC** can also provide the ability to connect to other components, e.g. Message Queues, which will help TiKV integrate with customers' data systems, and further extend the applicable scenarios of TiKV.
+
+![rawkv-cdc](../media/rawkv-cdc.png)
+
+### 2. Add Timestamp to Data
+
+As a kind of [Change Data Capture](https://en.wikipedia.org/wiki/Change_data_capture), timestamp (or version) is necessary to indicate which and when data is changed.
+
+#### 2.1 Requirement
+
+Among requests of a key, the timestamp must be monotonic with the sequence of data flushed to disk in TiKV.
+
+In general, if request `a` ["happened before"](https://en.wikipedia.org/wiki/Happened-before) `b`, then `Timestamp(a) < Timestamp(b)`. As to cross cluster replication, we provide [Causal Consistency](https://en.wikipedia.org/wiki/Causal_consistency) by keeping order of the timestamp the same as sequence of data flush to disk in TiKV. Downstream systems apply data according to the timestamp order, will always lead to the consistent result.
+
+At the same time, as RawKV doesn't provide cross-rows transaction and snapshot isolation, we treat requests of different keys as concurrent, to reduce complexity and improve performance.
+
+#### 2.2 Timestamp Generation
+
+Timestamp is generated by TiKV internally, to get a better overall performance and client compatibility.
+
+We use [HLC](https://cse.buffalo.edu/tech-reports/2014-04.pdf) method to generate timestamp, which not only is easy to use (by being closed to real world time), but also can capture causality.
+
+##### 2.2.1 Physical Part of Timestamp
+
+Physical part of HLC is acquired from TSO of [PD](https://github.com/tikv/pd).
+
+TSO is a global monotonically increasing timestamp, which help the generation of timestamp be independent to local clock of machine, and be immune to issues such as reverse between machine reboot.
+
+Physical part is refreshed by repeatedly acquiring TSO in a period of `500ms`, to keep it being closed the real world time.
+
+On startup, physical part must be initialized by a successful TSO, to ensure that HLC is larger than last running. RawKV write operations will fail if the initialization has not been completed. (TiKV also can't startup when PD is out of service).
+
+After that, HLC can tolerate fault of TSO. RawKV operations are normal, but time-related metrics such as RPO is not reasonable. Meanwhile, as TiKV-CDC being highly dependent to PD (for cluster and task management, checkpoint, etc.), the replication will not be working.
+
+##### 2.2.2 Logical Part of Timestamp
+
+Logical part is advanced on every write request.
+
+Moreover, the logical part is advanced on *Leader Transfer*. As TiKV is a distributed system, every store in the TiKV cluster has an instance of timestamp generator. As every key exists in only one region and only one peer (the leader) in the region can write, it's safe to generate timestamps locally in each store, except for leader transfer.
```

Comment:
> the logical part is advanced on *Leader Transfer*

when assigning a timestamp for a key-value put by a client, the physical part is generated by the local store's machine time, but how to get the logical part? I'm a little confused about this sentence, could you explain with some examples?

Reply:
I add a diagram and some words for this. PTAL~